File: claude_proxy_server.mjs
#!/usr/bin/env node

/**
 * Claude Code Node.js SDK Proxy Server with True Streaming Support
 * 
 * Maintains a single continuous session throughout the entire conversation.
 * All messages flow through the same query stream without creating new queries.
 * 
 * Communication Protocol:
 * - Input: JSONL messages via stdin (from Gromozeka)
 * - Output: JSONL messages via stdout (to Gromozeka) - REAL-TIME STREAMING
 * - Errors: Logged to stderr for debugging
 */

import readline from 'readline';
import { execSync } from 'child_process';
import path from 'path';
import fs from 'fs';

// Dynamically find Claude Code installation
function findClaudeCodeSdk() {
    try {
        const claudePath = execSync('which claude', { encoding: 'utf8' }).trim();
        console.error('[NodeSDK Proxy] Found claude at:', claudePath);
        
        const realPath = fs.realpathSync(claudePath);
        console.error('[NodeSDK Proxy] Resolved to:', realPath);
        
        const claudeDir = path.dirname(realPath);
        const sdkPath = path.join(claudeDir, 'sdk.mjs');
        
        if (fs.existsSync(sdkPath)) {
            console.error('[NodeSDK Proxy] Found SDK at:', sdkPath);
            return sdkPath;
        }
        
        const altSdkPath = realPath.replace('/bin/claude', '/lib/node_modules/@anthropic-ai/claude-code/sdk.mjs');
        if (fs.existsSync(altSdkPath)) {
            console.error('[NodeSDK Proxy] Found SDK at:', altSdkPath);
            return altSdkPath;
        }
        
        throw new Error(`SDK not found at expected locations: ${sdkPath} or ${altSdkPath}`);
    } catch (error) {
        console.error('[NodeSDK Proxy] Error finding Claude Code:', error.message);
        throw error;
    }
}

let query;
try {
    const sdkPath = findClaudeCodeSdk();
    const sdkModule = await import(sdkPath);
    query = sdkModule.query;
    console.error('[NodeSDK Proxy] Successfully loaded SDK');
} catch (error) {
    console.error('[NodeSDK Proxy] Failed to load Claude Code SDK:', error.message);
    console.error('[NodeSDK Proxy] Make sure Claude Code is installed: npm install -g @anthropic-ai/claude-code');
    process.exit(1);
}

// Configuration from environment variables
const config = {
    systemPrompt: process.env.CLAUDE_SYSTEM_PROMPT || '',
    permissionMode: process.env.CLAUDE_PERMISSION_MODE || 'default',
    model: process.env.CLAUDE_MODEL || undefined,
    resumeSession: process.env.CLAUDE_RESUME_SESSION || undefined,
    cwd: process.cwd()
};

console.error('[NodeSDK Proxy] Starting with config:', {
    ...config,
    systemPrompt: config.systemPrompt ? `<${config.systemPrompt.length} chars>` : 'none'
});

// Create readline interface for stdin
const rl = readline.createInterface({
    input: process.stdin,
    output: null,
    terminal: false
});

// Session and stream management
let activeQuery = null;
let sessionId = null;
let streamController = null;
let messageQueue = [];
let waitingForMessage = null;
let streamEnded = false;

/**
 * Send JSONL message to stdout (streaming)
 */
function sendMessage(message) {
    const json = JSON.stringify(message);
    console.log(json);
}

/**
 * Log to stderr for debugging
 */
function logError(...args) {
    console.error('[NodeSDK Proxy]', ...args);
}

/**
 * Create a continuous async iterator for streaming input to SDK
 * This iterator will keep yielding messages as they arrive from stdin
 * and only complete when the process is terminated
 */
async function* createContinuousInputStream() {
    logError('Creating continuous input stream');
    
    while (!streamEnded) {
        // Check if we have messages in queue
        if (messageQueue.length > 0) {
            const message = messageQueue.shift();
            logError('Yielding queued message to SDK');
            yield message;
        } else {
            // Wait for next message
            logError('Waiting for next message...');
            const message = await new Promise(resolve => {
                waitingForMessage = resolve;
            });
            
            if (message) {
                logError('Yielding new message to SDK');
                yield message;
            }
        }
    }
    
    logError('Input stream ended');
}

/**
 * Start the single continuous streaming query
 */
async function startContinuousStreamingQuery() {
    try {
        logError('Starting continuous streaming query');
        
        // Create abort controller for this session
        streamController = new AbortController();
        
        // Configure SDK options
        const options = {
            appendSystemPrompt: config.systemPrompt,
            permissionMode: config.permissionMode,
            cwd: config.cwd,
            model: config.model,
            resume: config.resumeSession,
            abortController: streamController
        };
        
        logError('Query options:', {
            ...options,
            appendSystemPrompt: options.appendSystemPrompt ? `<${options.appendSystemPrompt.length} chars>` : 'none'
        });
        
        // Start the query with continuous streaming input
        // This creates a SINGLE query that will handle ALL messages
        activeQuery = query({
            prompt: createContinuousInputStream(),
            options
        });
        
        // Process SDK messages as they arrive (REAL-TIME STREAMING)
        // This loop will continue for the entire conversation
        for await (const sdkMessage of activeQuery) {
            logError('SDK message (streaming):', sdkMessage.type);
            
            // Convert and send message immediately
            const streamMessage = convertSdkToStreamMessage(sdkMessage);
            if (streamMessage) {
                sendMessage(streamMessage);
            }
        }
        
        logError('Query stream ended');
        
    } catch (error) {
        logError('Error in continuous streaming query:', error);
        
        // Send error message
        sendMessage({
            type: 'system',
            subtype: 'error',
            data: {
                error: error.message || 'Unknown error'
            }
        });
    }
}

/**
 * Convert SDK message format to stream-json format
 */
function convertSdkToStreamMessage(sdkMessage) {
    switch (sdkMessage.type) {
        case 'system':
            if (sdkMessage.subtype === 'init') {
                sessionId = sdkMessage.session_id;
                logError('Session initialized:', sessionId);
                
                return {
                    type: 'system',
                    subtype: 'init',
                    data: {
                        session_id: sessionId,
                        model: sdkMessage.model,
                        tools: sdkMessage.tools || [],
                        mcp_servers: sdkMessage.mcp_servers || [],
                        permission_mode: sdkMessage.permissionMode
                    },
                    session_id: sessionId
                };
            }
            break;

        case 'assistant':
            // Stream assistant message immediately
            return {
                type: 'assistant',
                message: sdkMessage.message,
                session_id: sessionId || sdkMessage.session_id,
                parent_tool_use_id: sdkMessage.parent_tool_use_id
            };

        case 'user':
            // Echo user messages (shouldn't happen in normal flow)
            return {
                type: 'user',
                message: sdkMessage.message,
                session_id: sessionId || sdkMessage.session_id,
                parent_tool_use_id: sdkMessage.parent_tool_use_id
            };

        case 'result':
            // Result messages in streaming mode
            // Don't treat result as end of conversation
            logError('Got result message but keeping stream alive');
            return {
                type: 'result',
                subtype: sdkMessage.subtype,
                duration_ms: sdkMessage.duration_ms,
                duration_api_ms: sdkMessage.duration_api_ms,
                is_error: sdkMessage.is_error,
                num_turns: sdkMessage.num_turns,
                total_cost_usd: sdkMessage.total_cost_usd,
                usage: sdkMessage.usage,
                result: sdkMessage.result,
                session_id: sessionId || sdkMessage.session_id
            };

        default:
            logError('Unknown SDK message type:', sdkMessage.type);
            return null;
    }
}

/**
 * Process incoming user message
 */
function processUserMessage(userMessage) {
    logError('Processing user message:', 
        userMessage.message?.content?.[0]?.text?.substring(0, 100) || '<empty>');
    
    // Convert to SDK format
    const sdkMessage = {
        type: 'user',
        message: {
            role: 'user',
            content: userMessage.message.content
        },
        parent_tool_use_id: userMessage.parent_tool_use_id || null,
        session_id: userMessage.session_id || sessionId
    };
    
    // If someone is waiting for a message, resolve immediately
    if (waitingForMessage) {
        logError('Resolving waiting promise with message');
        const resolve = waitingForMessage;
        waitingForMessage = null;
        resolve(sdkMessage);
    } else {
        // Otherwise queue it
        logError('Queueing message');
        messageQueue.push(sdkMessage);
    }
}

/**
 * Process control request (interrupt, etc.)
 */
async function processControlRequest(controlRequest) {
    try {
        logError('Processing control request:', controlRequest.request?.subtype);
        
        if (controlRequest.request?.subtype === 'interrupt') {
            if (activeQuery && typeof activeQuery.interrupt === 'function') {
                await activeQuery.interrupt();
                
                sendMessage({
                    type: 'control_response',
                    response: {
                        request_id: controlRequest.request_id,
                        subtype: 'success'
                    }
                });
            } else {
                sendMessage({
                    type: 'control_response',
                    response: {
                        request_id: controlRequest.request_id,
                        subtype: 'error',
                        error: 'No active query to interrupt'
                    }
                });
            }
        }
    } catch (error) {
        logError('Error processing control request:', error);
        
        sendMessage({
            type: 'control_response',
            response: {
                request_id: controlRequest.request_id,
                subtype: 'error',
                error: error.message || 'Unknown error'
            }
        });
    }
}

// Start the continuous streaming query ONCE
startContinuousStreamingQuery();

// Handle incoming messages from stdin
rl.on('line', async (line) => {
    try {
        const message = JSON.parse(line);
        logError('Received message type:', message.type);
        
        if (message.type === 'user') {
            processUserMessage(message);
        } else if (message.type === 'control_request') {
            await processControlRequest(message);
        }
        
    } catch (error) {
        logError('Failed to parse input:', error);
    }
});

// Handle stdin close
rl.on('close', () => {
    logError('Stdin closed, ending stream...');
    streamEnded = true;
    
    // If waiting for message, resolve with null to end the stream
    if (waitingForMessage) {
        waitingForMessage(null);
    }
    
    if (streamController && !streamController.signal.aborted) {
        streamController.abort();
    }
    
    setTimeout(() => process.exit(0), 100);
});

// Handle process signals
process.on('SIGTERM', () => {
    logError('Received SIGTERM, exiting...');
    streamEnded = true;
    if (streamController && !streamController.signal.aborted) {
        streamController.abort();
    }
    process.exit(0);
});

process.on('SIGINT', () => {
    logError('Received SIGINT, exiting...');
    streamEnded = true;
    if (streamController && !streamController.signal.aborted) {
        streamController.abort();
    }
    process.exit(0);
});

logError('Node.js SDK proxy server started (continuous streaming mode)');

File: claude_proxy_server.py
#!/usr/bin/env python3
"""
Claude Code Proxy Server - обертка над Python SDK для Gromozeka
Принимает JSON через stdin/stdout и проксирует в Claude Code SDK
"""

import asyncio
import json
import os
import sys
import uuid
from typing import AsyncIterator, Any, Dict, Optional
from claude_code_sdk import (
    ClaudeSDKClient,
    ClaudeCodeOptions,
    UserMessage,
    AssistantMessage,
    SystemMessage,
    ResultMessage,
    TextBlock,
    ToolUseBlock,
    ToolResultBlock,
    ThinkingBlock,
)

class ClaudeProxy:
    def __init__(self):
        self.client = None
        self.session_id = str(uuid.uuid4())
        self.active_session_id = None
        self.request_counter = 0
        
    async def start(self):
        """Initialize Claude SDK client"""
        # Build options from environment variables
        options = ClaudeCodeOptions(
            permission_mode=os.environ.get("CLAUDE_PERMISSION_MODE", "acceptEdits"),
            system_prompt=os.environ.get("CLAUDE_SYSTEM_PROMPT"),
            model=os.environ.get("CLAUDE_MODEL"),
            resume=os.environ.get("CLAUDE_RESUME_SESSION"),
            max_thinking_tokens=10000,
        )
        
        self.client = ClaudeSDKClient(options)
        await self.client.connect()
        
        # Start receiving task
        asyncio.create_task(self.receive_from_claude())
        
    async def receive_from_claude(self):
        """Continuously receive messages from Claude"""
        try:
            async for message in self.client.receive_messages():
                # Convert SDK message to stream-json format
                output = self.convert_to_stream_json(message)
                if output:
                    print(json.dumps(output, ensure_ascii=False), flush=True)
        except asyncio.CancelledError:
            # Normal cancellation, don't print error
            pass
        except Exception as e:
            error_msg = {
                "type": "system",
                "subtype": "error",
                "data": {"error": str(e)}
            }
            print(json.dumps(error_msg, ensure_ascii=False), flush=True)
    
    def convert_to_stream_json(self, message) -> Optional[Dict[str, Any]]:
        """Convert SDK message to stream-json format compatible with Gromozeka"""
        try:
            # Convert different message types
            if isinstance(message, UserMessage):
                # Convert user message
                content = self._convert_content_blocks(message.content)
                return {
                    "type": "user",
                    "message": {
                        "role": "user",
                        "content": content
                    },
                    "session_id": self.active_session_id
                }
                
            elif isinstance(message, AssistantMessage):
                # Convert assistant message
                content = self._convert_content_blocks(message.content)
                return {
                    "type": "assistant",
                    "message": {
                        "role": "assistant",
                        "content": content,
                        "model": message.model
                    },
                    "session_id": self.active_session_id
                }
                
            elif isinstance(message, SystemMessage):
                # System messages - handle init and other subtypes
                if message.subtype == "init":
                    # Extract session ID from system message if available
                    if "session_id" in message.data:
                        self.active_session_id = message.data["session_id"]
                    elif not self.active_session_id:
                        self.active_session_id = self.session_id
                        
                return {
                    "type": "system",
                    "subtype": message.subtype,
                    "data": message.data,
                    "session_id": self.active_session_id
                }
                
            elif isinstance(message, ResultMessage):
                # Result message with usage stats - fields at top level for compatibility
                return {
                    "type": "result",
                    "subtype": message.subtype,
                    "duration_ms": message.duration_ms,
                    "duration_api_ms": message.duration_api_ms,
                    "is_error": message.is_error,
                    "num_turns": message.num_turns,
                    "session_id": message.session_id or self.active_session_id,
                    "total_cost_usd": message.total_cost_usd,
                    "usage": message.usage,
                    "result": message.result
                }
            else:
                # Unknown message type - log for debugging
                return {
                    "type": "system",
                    "subtype": "unknown_message",
                    "data": {"message_type": type(message).__name__, "content": str(message)}
                }
                
        except Exception as e:
            return {
                "type": "system",
                "subtype": "conversion_error", 
                "data": {"error": str(e), "message_type": type(message).__name__}
            }
    
    def _convert_content_blocks(self, content) -> list:
        """Convert SDK content blocks to stream-json format"""
        if isinstance(content, str):
            # Simple string content
            return [{"type": "text", "text": content}]
        
        if isinstance(content, list):
            # List of content blocks
            result = []
            for block in content:
                if isinstance(block, TextBlock):
                    result.append({"type": "text", "text": block.text})
                elif isinstance(block, ToolUseBlock):
                    result.append({
                        "type": "tool_use",
                        "id": block.id,
                        "name": block.name,
                        "input": block.input
                    })
                elif isinstance(block, ToolResultBlock):
                    result.append({
                        "type": "tool_result",
                        "tool_use_id": block.tool_use_id,
                        "content": block.content,
                        "is_error": block.is_error
                    })
                elif isinstance(block, ThinkingBlock):
                    result.append({
                        "type": "thinking",
                        "thinking": block.thinking,
                        "signature": block.signature
                    })
            return result
        
        # Fallback - convert to string
        return [{"type": "text", "text": str(content)}]
    
    async def handle_input(self, line: str):
        """Handle input from Gromozeka"""
        try:
            data = json.loads(line)
            
            if data.get("type") == "user":
                # Extract message content and session ID
                message_data = data["message"]
                content_blocks = message_data.get("content", [])
                session_id = data.get("session_id", self.session_id)
                
                # Store active session ID
                self.active_session_id = session_id
                
                # Convert content blocks to text (SDK expects string for simple queries)
                text_content = ""
                for block in content_blocks:
                    if isinstance(block, dict) and block.get("type") == "text":
                        text_content += block.get("text", "")
                
                # Send to Claude via SDK
                await self.client.query(text_content, session_id=session_id)
                
            elif data.get("type") == "control_request":
                # Handle control messages
                request_id = data.get("request_id")
                request = data.get("request", {})
                subtype = request.get("subtype")
                
                if subtype == "interrupt":
                    # Send interrupt to SDK
                    await self.client.interrupt()
                    
                    # Send control response
                    response = {
                        "type": "control_response",
                        "response": {
                            "request_id": request_id,
                            "subtype": "interrupt",
                            "error": None
                        }
                    }
                    print(json.dumps(response, ensure_ascii=False), flush=True)
                    
        except json.JSONDecodeError as e:
            error_msg = {
                "type": "system",
                "subtype": "parse_error",
                "data": {"error": f"Invalid JSON: {e}", "line": line}
            }
            print(json.dumps(error_msg, ensure_ascii=False), flush=True)
        except Exception as e:
            error_msg = {
                "type": "system",
                "subtype": "error",
                "data": {"error": str(e)}
            }
            print(json.dumps(error_msg, ensure_ascii=False), flush=True)
    
    async def run(self):
        """Main loop - read from stdin, write to stdout"""
        await self.start()
        
        # Don't send our own init - Claude SDK will send one
        # Just store initial session ID
        self.active_session_id = self.session_id
        
        # Read from stdin
        loop = asyncio.get_event_loop()
        reader = asyncio.StreamReader()
        protocol = asyncio.StreamReaderProtocol(reader)
        await loop.connect_read_pipe(lambda: protocol, sys.stdin)
        
        while True:
            try:
                line_bytes = await reader.readline()
                if not line_bytes:
                    break
                line = line_bytes.decode('utf-8').strip()
                if line:
                    await self.handle_input(line)
            except asyncio.CancelledError:
                break
            except Exception as e:
                error_msg = {
                    "type": "system",
                    "subtype": "error",
                    "data": {"error": str(e)}
                }
                print(json.dumps(error_msg, ensure_ascii=False), flush=True)
    
    async def cleanup(self):
        """Clean up resources"""
        if self.client:
            try:
                await self.client.disconnect()
            except Exception as e:
                # Log but don't fail on cleanup errors
                sys.stderr.write(f"Cleanup error: {e}\n")

async def main():
    """Main entry point"""
    proxy = ClaudeProxy()
    try:
        await proxy.run()
    except KeyboardInterrupt:
        sys.stderr.write("Interrupted by user\n")
    except Exception as e:
        sys.stderr.write(f"Fatal error: {e}\n")
        raise
    finally:
        await proxy.cleanup()

if __name__ == "__main__":
    # Set up event loop and run
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception:
        sys.exit(1)